designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,311 @@
+
+For example, with the rule [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports), one might want to test that the rule schema fails validation when passed:
+
+- Something that isn't an array
+- An empty array
+- An array containing an item that isn't a string nor object
+- An array containing an object that is missing required properties like `name`, or has unknown properties
+- Any other invalid combinations of input
+
+Note that the goal is not to test that [ajv](https://github.com/ajv-validator/ajv) (which performs the validation using JSON Schema) itself works properly, as we can treat that third-party dependency as a blackbox, but instead to test that the rule author has written out their schema correctly.
+
+It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when no automated testing exists around this today. Schemas are often too lenient, allowing extra input that the rule implementation ignores or doesn't handle properly. This can result in rules silently misbehaving, not respecting the wishes or intentions of the user's configuration, or just crashing.
+
+By enabling testing of schemas, we'll make it easier to write higher-quality schemas and rules, thus improving the user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+Add support to the rule tester for a new `fatal` test case array in addition to `valid` and `invalid`.
+
+Example with a schema validation error because a non-existent option is passed:
+
+```json
+{
+    "fatal": [
+        {
+            "code": "foo();",
+            "options": [{ "checkFoo": true, "nonExistentOption": true }],
+            "message": "Value {\"checkFoo\":true,\"nonExistentOption\":true} should NOT have additional properties."
+        }
+    ]
+}
+```
+
+To test a custom exception thrown by a rule, simply use the exception text as the message: `"message": "Some custom exception message from the rule."`.
+
+A convenience feature to go along with this is to allow an alternative property `errorType: 'schema' | 'rule'` to be specified indicating the type of exception without indicating the exception text.
+This would be useful since most of the time, the user may not care about the exact text of the exception, but only that the particular type of exception was thrown.
+This also makes it easier to update the exception text without having to update the messages in all the test cases.
+Alternate names for this property to consider: `messageErrorType`, `messageType`, `messageId`, `source`.
```

Comment:
> The testing framework forces you to provide the expected error message (for configuration but for all the other errors as well). Not requiring so would create the possibility that the test is failing for an different or unexpected reason. So I would recommend forcing to give the expected error message (and have the test failure make that easy to copy/paste).
> 
> I don't know how I feel about a still helpful but less precise "CUSTOM_RULE_ERROR". If the error message will be shown to the user and is controllable by the rule author (you could catch the error and re-throw it with a nicer error message), then I would really recommend that authors have to provide it. Being able to see it in the test source code means you can see the error that will be presented to the user, and figure out how it could be better presented or made more understandable.
> 
> For SCHEMA_VALIDATION_ERROR, I imagine the rule author doesn't control the error message, but forcing the error message instead of using this generic id might catch some problems where you are getting an different/unexpected validation error. For instance if the schema changes and you forget to update one of the tests.

@jfmengels I appreciate your feedback and want to focus in on this part of your comment regarding my proposed convenience feature where the user could specify the type of error instead of the actual exception text.

I do agree that it can sometimes be helpful to include the exact exception text in the test case. However, I still currently lean towards including this feature, since I think knowing that an exception occurred during "schema validation" or "rule implementation" is likely to be sufficient most of the time, and including lengthy exception text can be a hassle (although using a regexp for `message` can help mitigate that).

The main downside I see is that a test using `errorType: 'rule'` (latest iteration of the property) wouldn't distinguish between the multiple kinds of exceptions that a rule implementation might throw.

One idea is that we could use the exception class name in the property instead. `errorType: 'SchemaValidationError' | 'SomeCustomRuleError' | 'Error'` where the schema validation code uses `throw new SchemaValidationError()` and the rule implementation uses `throw new SomeCustomRuleError()` or even just `throw new Error`. This could provide a shorthand for distinguishing between different types of exceptions without having to provide the full message text. I think it's potentially more elegant and useful than just `errorType: 'schema' | 'rule'`.

So I'm interested to hear what everyone thinks about (1) including the convenience feature, and (2) using the exception class name as a way to differentiate the source of the exception.